### ISSUE TITLE
If you Can realize LPRNet using DL4J

### ISSUE THREAD

**Issue by mlistbox** [deeplearning4j/deeplearning4j#9874]
#### Issue Description

Please describe our issue, along with:
- expected behavior
- encountered behavior

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version
* Platform information (OS, etc)
* CUDA version, if used
* NVIDIA driver version, if in use

#### Additional Information

Where applicable, please also provide:

* Full log or exception stack trace (ideally in a Gist: gist.github.com)
* pom.xml file or similar (also in a Gist)

#### Contributing

If you'd like to help us fix the issue by contributing some code, but would
like guidance or help in doing so, please mention it!

**Comment by mlistbox** [comment]
Issue Description
realize LPRNet using DL4J but throw exception.DL4JInvalidInputException: Cannot do forward pass in Convolution layer (layer name = sbb_21, layer index = 15): input array channels does not match CNN layer configuration (data format = NCHW, data input channels = 64, [minibatch, channels, height, width]=[8, 64, 18, 44]; expected input channels = 128) (layer name: sbb_21, layer index: 15, layer type: ConvolutionLayer).

expected behavior
Can  finish train.

encountered behavior
throw exception.DL4JInvalidInputException: Cannot do forward pass in Convolution layer (layer name = sbb_21, layer index = 15): input array channels does not match CNN layer configuration (data format = NCHW, data input channels = 64, [minibatch, channels, height, width]=[8, 64, 18, 44]; expected input channels = 128) (layer name: sbb_21, layer index: 15, layer type: ConvolutionLayer).

Version Information
1.0.0-M2.1

<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>com.kt</groupId>
    <artifactId>dl4j_lpr</artifactId>
    <version>1.0-SNAPSHOT</version>
    <properties>
        <maven.compiler.source>14</maven.compiler.source>
        <maven.compiler.target>14</maven.compiler.target>
        <java.version>1.8</java.version>
    </properties>
    <dependencies>
    <dependency>
        <groupId>org.deeplearning4j</groupId>
        <artifactId>deeplearning4j-core</artifactId>
        <version>1.0.0-M2.1</version>
    </dependency>
    <dependency>
        <groupId>org.deeplearning4j</groupId>
        <artifactId>deeplearning4j-zoo</artifactId>
        <version>1.0.0-M2.1</version>
    </dependency>
    <dependency>
        <groupId>org.nd4j</groupId>
        <artifactId>nd4j-api</artifactId>
        <version>1.0.0-M2.1</version>
    </dependency>
    <dependency>
        <groupId>org.nd4j</groupId>
        <artifactId>nd4j-native</artifactId>
        <version>1.0.0-M2.1</version>
    </dependency>
    <dependency>
        <groupId>org.nd4j</groupId>
        <artifactId>nd4j-native-platform</artifactId>
        <version>1.0.0-M2.1</version>
    </dependency>
        <dependency>
            <groupId>org.bytedeco</groupId>
            <artifactId>opencv</artifactId>
            <version>4.5.1-1.5.5</version>
        </dependency>
    <dependency>
        <groupId>org.bytedeco</groupId>
        <artifactId>opencv-platform</artifactId>
        <version>4.5.1-1.5.5</version>
    </dependency>
    <dependency>
        <groupId>org.bytedeco</groupId>
        <artifactId>javacv</artifactId>
        <version>1.5.5</version>
    </dependency>
        <dependency>
        <groupId>org.bytedeco</groupId>
        <artifactId>javacv-platform</artifactId>
        <version>1.5.5</version>
    </dependency>
        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-log4j12</artifactId>
            <version>1.7.21</version>
        </dependency>
        <dependency>
            <groupId>log4j</groupId>
            <artifactId>log4j</artifactId>
            <version>1.2.17</version>
        </dependency>
        <dependency>
            <groupId>org.projectlombok</groupId>
            <artifactId>lombok</artifactId>
            <version>1.18.24</version>
            <scope>provided</scope>
        </dependency>
</dependencies>
<build>
    <plugins>
        
        <plugin>
            <artifactId>maven-source-plugin</artifactId>
            <version>3.0.1</version>
            <configuration>
                <attach>true</attach>
            </configuration>
            <executions>
                <execution>
                    <phase>compile</phase>
                    <goals>
                        <goal>jar</goal>
                    </goals>
                </execution>
            </executions>
        </plugin>
        <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-compiler-plugin</artifactId>
            <version>3.1</version>
            <configuration>
                <source>1.8</source>
                <target>1.8</target>
            </configuration>
        </plugin>
    </plugins>
</build>
</project>


@Slf4j
public class LprNet {


    public static void detection() throws IOException {
        int width = 94;
        int height = 24;
        int nChannels = 3;
        int batchSize = 8;
        int nEpochs = 10;
        double learningRate = 1e-3;
        double lrMomentum = 0.9;
        int seed = 1000;
        Random rng = new Random(seed);
        RandomPathFilter pathFilter = new RandomPathFilter(rng);
        String modelFilename = "LprNetModel.zip";
            /*
        {
            @Override
            protected boolean accept(String name) {
                //name = "/test/";
                try {
                    return new File(new URI(name)).exists();
                } catch (URISyntaxException ex) {
                    throw new RuntimeException(ex);
                }
            }
        };
        */
        String dataDir = new ClassPathResource("/").getFile().getPath();
        File imageDir = new File(dataDir, "testjpg");
        log.info("Load LPRNet train data...");
        //log.info(Arrays.toString(PlateLableGenerator.CHARS));
        InputSplit[] data = new FileSplit(imageDir, NativeImageLoader.ALLOWED_FORMATS, rng).sample(pathFilter, 0.8, 0.2);
        InputSplit trainData = data[0];
        //trainData.
        //List<String> trainlables=new ArrayList<>();
        //Iterator<String> it=trainData.locationsPathIterator();
        //while(it.hasNext())
        {
            //trainlables.add(FilenameUtils.getBaseName(it.next().toString()));
        }
        InputSplit testData = data[1];
        //List<String> testlables=new ArrayList<>();
        //it=testData.locationsPathIterator();
        //while(it.hasNext())
        {
            //testlables.add(FilenameUtils.getBaseName(it.next().toString()));
        }
        //ImageTransform transform = new MultiImageTransform(randNumGen, new ShowImageTransform("Display - before "));
        PlateLableGenerator labelMaker=new PlateLableGenerator();
        //ParentPathLabelGenerator labelMaker = new ParentPathLabelGenerator();

        ImageRecordReader recordReaderTrain = new ImageRecordReader(height, width, nChannels, labelMaker);
        //ImageTransform transform = new MultiImageTransform(rng, new ShowImageTransform("Display - before "));
        recordReaderTrain.initialize(trainData);
        //List<Writable> lw = recordReaderTrain.next();
        // then lw[0] =  NDArray shaped [1,3,50,50] (1, channels, height, width)
        //recordReaderTrain.setLabels(trainlables);

        ImageRecordReader recordReaderTest = new ImageRecordReader(height, width, nChannels,
                labelMaker);
        recordReaderTest.initialize(testData);
        //recordReaderTest.setLabels(testlables);
        //int outputNum = recordReaderTrain.numLabels();
        RecordReaderDataSetIterator train = new RecordReaderDataSetIterator(recordReaderTrain, batchSize, 1,1, true);

        train.setPreProcessor(new ImagePreProcessingScaler(0, 1));

        RecordReaderDataSetIterator test = new RecordReaderDataSetIterator(recordReaderTest, batchSize, 1,1, true);
        test.setPreProcessor(new ImagePreProcessingScaler(0, 1));

        ComputationGraphConfiguration conf =
                new NeuralNetConfiguration.Builder().seed(seed)
                        .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT).gradientNormalization(GradientNormalization.RenormalizeL2PerLayer)
                        .gradientNormalizationThreshold(1.0).updater(new Adam.Builder().learningRate(learningRate).build())
                        .updater(new Nesterovs.Builder().learningRate(learningRate).momentum(lrMomentum).build())
                        .activation(Activation.RELU)
                        .trainingWorkspaceMode(WorkspaceMode.ENABLED).inferenceWorkspaceMode(WorkspaceMode.ENABLED)
                        .graphBuilder()
                        .addInputs("in")
                        .layer("conv1", new Convolution2D.Builder().kernelSize(3, 3).stride(1,1)
                                .nIn(3).nOut(64).build(),"in")
                        .layer("BN2d_1",new BatchNormalization.Builder().nOut(64).build(),"conv1")
                        .layer("active_1", new ActivationLayer.Builder().activation(Activation.RELU).build(), "BN2d_1")
                        //.layer("MaxPool3d", new Subsampling3DLayer.Builder(Subsampling3DLayer.PoolingType.MAX).kernelSize(1,3, 3).stride(1, 1,1)
                        //       .build(),"active_1")

                        .layer("MP2d_1", new SubsamplingLayer.Builder(PoolingType.MAX).kernelSize(3, 3).stride( 1,1).build(),"active_1")
                        // .layer("MP3d_1", new Subsampling3DLayer.Builder().kernelSize(1, 3,3).poolingType(Subsampling3DLayer.PoolingType.MAX).stride(1, 1,1).build(),"active_1")
                        .layer("sbb_11", new ConvolutionLayer.Builder().kernelSize(1, 1).nIn(64).nOut(128).build(),"MP2d_1")
                        .layer("sbb_12", new ConvolutionLayer.Builder().kernelSize(3, 1).padding(1,0).nIn(128).nOut(128).build(),"sbb_11")
                        .layer("sbb_13", new ConvolutionLayer.Builder().kernelSize(1, 3).padding(0,1).nIn(128).nOut(128).build(),"sbb_12")
                        .layer("sbb_14", new ConvolutionLayer.Builder().kernelSize(1, 1).nIn(128).nOut(128).build(),"sbb_13")

                        .layer("BN2d_2",new BatchNormalization.Builder().nOut(128).build(),"sbb_14")
                        .layer("active_2", new ActivationLayer.Builder().activation(Activation.RELU).build(), "BN2d_2")

                        //.layer("MP3d_2", new Subsampling3DLayer.Builder().kernelSize(1, 3,3).poolingType(Subsampling3DLayer.PoolingType.MAX).stride(2, 1,2).build(),"active_2")

                        .layer("MP2d_2", new SubsamplingLayer.Builder(PoolingType.MAX).kernelSize(3, 3).stride( 1,2).build(),"active_2")
                        //.addVertex("Stack2", new StackVertex(), "MP2d_2")
                        .layer("MP2d_ts1",new TranSposeFor3Dto2D(0,2,3,1),"MP2d_2")
                        //.layer("MP2d_ts1",new TranSpose
                        .layer("MP2d_3", new SubsamplingLayer.Builder(PoolingType.MAX).kernelSize(1, 1).stride( 1,2).build(),"MP2d_ts1")
                        .layer("MP2d_ts2",new TranSposeFor3Dto2D(0,3,1,2),"MP2d_3")

                        .layer("sbb_21", new ConvolutionLayer.Builder().kernelSize(1, 1).nIn(64).nOut(256).build(),"MP2d_ts2")
                        .layer("sbb_22", new ConvolutionLayer.Builder().kernelSize(3, 1).padding(1,0).nIn(256).nOut(256).build(),"sbb_21")
                        .layer("sbb_23", new ConvolutionLayer.Builder().kernelSize(1, 3).padding(0,1).nIn(256).nOut(256).build(),"sbb_22")
                        .layer("sbb_24", new ConvolutionLayer.Builder().kernelSize(1, 1).nIn(256).nOut(256).build(),"sbb_23")

                        .layer("BN2d_3",new BatchNormalization.Builder().nOut(256).build(),"sbb_24")
                        .layer("active_3", new ActivationLayer.Builder().activation(Activation.RELU).build(), "BN2d_3")

                        .layer("sbb_31", new ConvolutionLayer.Builder().kernelSize(1, 1).nIn(256).nOut(256).build(),"active_3")
                        .layer("sbb_32", new ConvolutionLayer.Builder().kernelSize(3, 1).padding(1,0).nIn(256).nOut(256).build(),"sbb_31")
                        .layer("sbb_33", new ConvolutionLayer.Builder().kernelSize(1, 3).padding(0,1).nIn(256).nOut(256).build(),"sbb_32")
                        .layer("sbb_34", new ConvolutionLayer.Builder().kernelSize(1, 1).nIn(256).nOut(256).build(),"sbb_33")

                        .layer("BN2d_4",new BatchNormalization.Builder().nOut(256).build(),"sbb_34")
                        .layer("active_4", new ActivationLayer.Builder().activation(Activation.RELU).build(), "BN2d_4")

                        .layer("MP2d_4", new SubsamplingLayer.Builder(PoolingType.MAX).kernelSize(3, 3).stride( 1,2)
                                .build(),"active_4")
                        .layer("MP2d_5", new SubsamplingLayer.Builder(PoolingType.MAX).kernelSize(1, 1).stride( 1,4)
                                .build(),"MP2d_4")

                        .layer("d1",new DropoutLayer.Builder(0.5).build(),"MP2d_5")

                        .layer("conv2", new ConvolutionLayer.Builder().kernelSize(4, 1).stride(1,1)
                                .nIn(64).nOut(256).build(),"d1")
                        .layer("BN2d_5",new BatchNormalization.Builder().nOut(256).build(),"conv2")
                        .layer("active_5", new ActivationLayer.Builder().activation(Activation.RELU).build(), "BN2d_5")

                        .layer("d2",new DropoutLayer.Builder(0.5).build(),"active_5")

                        .layer("conv3", new ConvolutionLayer.Builder().kernelSize(13, 1).stride(1,1)
                                .nIn(256).nOut(68).build(),"d2")
                        .layer("BN2d_6",new BatchNormalization.Builder().nOut(68).build(),"conv3")
                        .layer("active_6", new ActivationLayer.Builder().activation(Activation.RELU).build(), "BN2d_6")
                        //.layer("conv4", new ConvolutionLayer.Builder().kernelSize(1, 1).stride(1,1).nIn(516).nOut(68).build(),"active_6")
                        .addLayer("output", new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                                .activation(Activation.SOFTMAX).nOut(800).build(), "active_6")
                        .setOutputs("output")
                        //.layer("output", new OutputLayer.Builder(
                        //        LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD).nIn(66).nOut(66)
                        //        .activation(Activation.SOFTMAX) // radial basis function required
                        //        .build(), "conv4")
                        //.setOutputs("output")
                        .setInputTypes(InputType.convolutionalFlat(24,94, 3))
                        .build();


        ComputationGraph model = new ComputationGraph(conf);
        model.init();

        log.info("LPRNET is:{}",model.summary(InputType.convolutional(height, width, nChannels)));
        log.info("Train model...");

        model.setListeners(new ScoreIterationListener(1));
        for (int i = 0; i < nEpochs; i++) {
            train.reset();
            while (train.hasNext()) {
                //DataSet ds = train.next();
                //System.out.println(ds);
                //log.info(ds);
                model.fit(train.next());
            }
            log.info("*** Completed epoch {} ***", i);
        }
        ModelSerializer.writeModel(model, modelFilename, true);
        //return network;
    }
    public static void main(String[] args) throws Exception {
        detection();
    }
}


@Slf4j
public class TranSposeFor3Dto2D extends SameDiffLambdaLayer {

    private int[] dimensions;

    /**
     *
     * @param dimensions Dimensions to calculate L2 norm over.
     *                   For DenseLayer/FeedForward input, this would be dimension 1
     *                   For RNNs, this would also be dimension 1 (to normalize each time step separately)
     *                   For CNNs, this would be dimensions 1, 2 and 3
     */
    public TranSposeFor3Dto2D(int... dimensions){
        this.dimensions = dimensions;
    }

    private TranSposeFor3Dto2D(){
        //Add a private no-arg constructor for use in JSON deserialization
    }

    @Override
    public SDVariable defineLayer(SameDiff sameDiff, SDVariable layerInput) {
        //Note the 'keepdims' arg: this will keep the dimensions, so we can auto-broadcast the later division
        //For example, if input is shape [3,4,5,6] and dimensions are [1,2,3] (i.e., for CNN activations)
        //then Norm2 has shape [3,1,1,1] - or if keepDims=false was used, it would have shape [3]
        //log.info(layerInput.shape().toString());
        SDVariable norm2 = layerInput.permute(dimensions);
        //log.info(norm2.shape().toString());
        return norm2;
    }

    //Getters and setters for JSON serialization
    public int[] getDimensions(){
        return dimensions;
    }

    public void setDimensions(int[] dimensions){
        this.dimensions = dimensions;
    }

    public InputType getOutputType(int layerIndex, InputType inputType) {
        return inputType;
    }

}

**Comment by mlistbox** [comment]
public class PlateLableGenerator implements PathLabelGenerator {
    public static String CHARS = "京沪津渝冀晋蒙辽吉黑苏浙皖闽赣鲁豫鄂湘粤桂琼川贵云藏陕甘青宁新0123456789ABCDEFGHJKLMNPQRSTUVWXYZIO-";
    @Override
    public Writable getLabelForPath(String path) {
        //String p=FilenameUtils.getBaseName(path);
        //int[] s=new int[p.length()];
        //for(int i=0;i<p.length();i++)
        //{
        //    s[i]=CHARS.indexOf(p.charAt(i));
        //}
        return new Text(FilenameUtils.getBaseName(path));
    }

    @Override
    public Writable getLabelForPath(URI uri) {
        return this.getLabelForPath((new File(uri)).toString());
    }

    @Override
    public boolean inferLabelClasses() {
        return true;
    }
}

**Comment by mlistbox** [comment]
InputTypeUtil.getPreProcessorForInputTypeCnn3DLayers(inputType, this.getLayerName());


public static InputPreProcessor getPreProcessorForInputTypeCnn3DLayers(InputType inputType, String layerName) {
        switch(inputType.getType()) {
        case FF:
            log.info("Automatic addition of FF -> CNN3D preprocessors: not yet implemented (layer name: \"" + layerName + "\")");
            return null;
        case RNN:
            log.warn("Automatic addition of RNN -> CNN3D preprocessors: not yet implemented (layer name: \"" + layerName + "\")");
            return null;
        case CNN3D:
            return null;
        default:
            throw new RuntimeException("Unknown input type: " + inputType);
        }
    }

This function does not support NCHW

**Comment by agibsonccc** [comment]
@mlistbox  if you want to try newer architectures using samediff go ahead. Beyond keras import though...we're not going to add too many features to the legacy api. I'm closing this due to that reason. Please reach out on the forums: https://community.konduit.ai/ if you want to try using the newer api for this. Thanks!
